ovs-p4rt/ovs_p4rt.cc

```diff
@@ -0,0 +1,414 @@
+// Copyright (c) 2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+// TODO: ovs-p4rt logging
+#include <string>
+#include <grpcpp/grpcpp.h>
+#include <arpa/inet.h>
+
+#include "absl/flags/flag.h"
+#include "p4/v1/p4runtime.grpc.pb.h"
+#include "openvswitch/ovs-p4rt.h"
+#include "p4_name_mapping.h"
+#include "ovs_p4rt_session.h"
+
+ABSL_FLAG(std::string, grpc_addr, "127.0.0.1:9339",
+          "P4Runtime server address.");
+ABSL_FLAG(uint64_t, device_id, 1, "P4Runtime device ID.");
+
+using OvsP4rtStream =
+    ::grpc::ClientReaderWriter<p4::v1::StreamMessageRequest,
+                               p4::v1::StreamMessageResponse>;
+
+std::string EncodeByteValue(int arg_count...) {
+  std::string byte_value;
+  va_list args;
+  va_start(args, arg_count);
+
+  for (int arg = 0; arg < arg_count; ++arg) {
+    uint8_t byte = va_arg(args, int);
+    byte_value.push_back(byte);
+  }
+
+  va_end(args);
+  return byte_value;
+}
+
+std::string CanonicalizeIp(const uint32_t ipv4addr) {
+  return EncodeByteValue(4, (ipv4addr & 0xff),
+                            ((ipv4addr >> 8) & 0xff),
+                            ((ipv4addr >> 16) & 0xff),
+                            ((ipv4addr >> 24) & 0xff));
+}
+
+std::string CanonicalizeMac(const uint8_t mac[6]) {
+  return EncodeByteValue(6, (mac[0] & 0xff),
+                            (mac[1] & 0xff),
+                            (mac[2] & 0xff),
+                            (mac[3] & 0xff),
+                            (mac[4] & 0xff),
+                            (mac[5] & 0xff));
+}
+
+int GetTableId(const ::p4::config::v1::P4Info &p4info,
+                 const std::string &t_name) {
+  for (const auto &table : p4info.tables()) {
+    const auto &pre = table.preamble();
+    if (pre.name() == t_name) return pre.id();
+  }
+  return -1;
```

Comment:
1st thing stratum does is to find table by this id. In case of -1, it returns "ERR_INVALID_P4_INFO".  All this is on the P4 Runtime server side. Do you think we should add a check as part of P4RT client?

Reply:
All the standard client implementations I have seen, sends the request to server and input validation is handled by server

Reply:
As long as you know that the server deals well with the invalid parameter and returns failure, there is no compelling need to do input validation yourself.